src/dbcp/transform/helpers.py

```diff
@@ -241,66 +257,66 @@ def parse_dates(series: pd.Series, expected_mean_year=2020) -> pd.Series:
         return multiformat_string_date_parser(series)
 
 
-def _geocode_row(
-    ser: pd.Series, client: GoogleGeocoder, state_col="state", locality_col="county"
-) -> List[str]:
-    """Function to pass into pandas df.apply() to geocode state/locality pairs.
+def _geocode_and_add_fips(
+    nan_fips: pd.DataFrame, state_col="state", locality_col="county", api="geocodio"
+) -> pd.DataFrame:
+    """Geocode locality names in a dataframe and add FIPS codes.
 
-    Args:
-        ser (pd.Series): a row of a larger dataframe to geocode
-        client (GoogleGeocoder): client for Google Maps Platform API
-        state_col (str, optional): name of the column of state names. Defaults to 'state'.
-        locality_col (str, optional): name of the column of locality names. Defaults to 'county'.
+    This is used for records that are not assigned a FIPS code by the initial lookup.
 
+    Args:
+        nan_fips: dataframe with state and locality columns
+        state_col: name of the state column
+        locality_col: name of the locality column
+        api: name of the geocoding API to use
     Returns:
-        List[str]: geocoded_locality_name, geocoded_locality_type, and geocoded_containing_county
+        dataframe with geocoded locality information
     """
-    client.geocode_request(name=ser[locality_col], state=ser[state_col])
-    return client.describe()
-
+    # Deduplicate on the state and locality columns to minimize API calls
+    key_cols = [state_col, locality_col]
+    deduped_nan_fips = nan_fips.loc[:, key_cols].drop_duplicates()
+    if api == "google":
+        geocoding_module = google_maps
+    elif api == "geocodio":
+        geocoding_module = geocodio
+    else:
+        raise ValueError(f"Unknown API: {api}")
 
-@GEOCODER_CACHE.cache()
-def _geocode_locality(
-    state_locality_df: pd.DataFrame, state_col="state", locality_col="county"
-) -> pd.DataFrame:
-    """Use Google Maps Platform API to look up information about state/locality pairs in a dataframe.
+    deduped_geocoded = geocoding_module._geocode_locality(
+        deduped_nan_fips,
+        # pass subset to _geocode_locality to maximize chance of a cache hit
+        # (this way other columns can change but caching still works)
+        state_col=state_col,
+        locality_col=locality_col,
+    )
 
-    Args:
-        state_locality_df (pd.DataFrame): dataframe with state and locality columns
-        state_col (str, optional): name of the column of state names. Defaults to 'state'.
-        locality_col (str, optional): name of the column of locality names. Defaults to 'county'.
+    # recombine deduped geocoded data with original nan_fips
+    geocoded_deduped_nan_fips = pd.concat(
```

Comment:
Is the reason we put the empty rows in while geocoding so that this concat can still work?

Reply:
Yes it's so we can preserve the index. Do you think it'd be cleaner to not add empty rows?

Reply:
Yeah, I think so, and intuitively seems like it would make the joining a little simpler. But it's not a blocking concern I think - refactoring would require (a) adding tests for this behavior and then (b) figuring out what sequence of joins etc. works for this use case, so it's not like it's a 5-minute dealio. Might be a good follow-up PR though.